### DISCUSSION TITLE
Serverless (Bref) deployment breaks most Livewire components

### DISCUSSION THREAD

**Discussion by selfsimilar** [livewire/livewire#5031] [1. Help] [answered]
My App leans heavily on Livewire, and runs like a champ locally on laptop under both Valet (Nginx, Redis cache, MariaDB) and Sail.

However, deploying to AWS Lambda seems to break Livewire in several ways.

### Break 1 - 419 Errors.

This affects all Livewire components. They render fine in the Blade context, so initial page loads look fine. But any interaction will generate a 419 error. The only way I've been able to fix this so far is to add `'livewire/*'` to the protected `$except` property of the `VerifyCsrfToken` middleware. This doesn't seem like a very safe change, so I don't plan for this to be the long-term fix, but it at least fixes all the 419 errors.

### Break 2 - Corrupt data when trying to hydrate the [foo] component.

Once the 419 errors are gone, some Livewire components begin to work, but most now throw the above exception. I haven't yet found what all the broken components have in common, but all the working components are all very simple with few properties that are untyped, or of very simple `string` or `int` type.

However, both of these errors are possibly related to session issues, and the fact that everything works fine locally makes me suspicious that this is at root a mis-configuration, not a widespread issue with poorly implementing the Livewire components.

So more details on my efforts so far:

Serverless configuration is using Bref to create the Lambda Layers, the Redis Elasticache, and the assets S3 bucket/Cloudfront. The VPC, subnets, and RDS MariaDB are created in Terraform and salient information is passed to Bref via Parameter Store.

I've also tried using both 'cookie' and 'redis' as `SESSION_DRIVER` values without improvement to the above errors.

As far as I can tell, everything else is working fine at this point except for these Livewire errors, so I'm pretty frustrated and scratching my head.

**Comment by selfsimilar** [comment]
Supporting details:


<details>
<summary>
`serverless.yml`
</summary>
<pre>
service: laravel

params:
  default:
    app_key: ${ssm:/laravel/app_key}
    cache_instance_size: cache.t3.micro
    certificate_arn: ${ssm:/laravel/certificate_arn}
    db_database: ${ssm:/laravel/db_name}
    db_host: ${ssm:/laravel/db_host}
    db_password: ${ssm:/laravel/db_pwd}
    db_port: ${ssm:/laravel/db_port}
    db_username: ${ssm:/laravel/db_user}
    debug: true
    private_route_table_1a_id: ${ssm:/laravel/private_route_table_1a_id}
    private_route_table_1b_id: ${ssm:/laravel/private_route_table_1b_id}
    private_subnet_1a_id: ${ssm:/laravel/private_subnet_1a_id}
    private_subnet_1b_id: ${ssm:/laravel/private_subnet_1b_id}
    private_subnet_db1_id: ${ssm:/laravel/private_subnet_db1_id}
    private_subnet_db2_id: ${ssm:/laravel/private_subnet_db2_id}
    public_route_table_id: ${ssm:/laravel/public_route_table_id}
    public_subnet_1a_id: ${ssm:/laravel/public_subnet_1a_id}
    public_subnet_1b_id: ${ssm:/laravel/public_subnet_1b_id}
    vpc_id: ${ssm:/laravel/vpc_id}
  dev:
    app_env: dev
    asset_bucket_name: dev.assets.example.net
    domain: dev.my.example.net
  prod:
    app_env: production
    asset_bucket_name: assets.example.net
    domain: my.example.net
    debug: false
  stage:
    app_env: stage
    asset_bucket_name: stage.assets.example.net
    domain: stage.my.example.net

provider:
  name: aws
  profile: terraform-stage
  # The AWS region in which to deploy (us-east-1 is the default)
  region: us-east-2
  # The stage of the application, e.g. dev, production, staging… ('dev' is the default)
  stage: stage
  runtime: provided.al2
  environment:
    NOOP: false
    APP_DEBUG: ${param:debug}
    APP_ENV: ${param:app_env}
    APP_KEY: ${param:app_key}
    APP_STORAGE: /tmp
    APP_URL: https://${param:domain}
    DB_DATABASE: ${param:db_database}
    DB_HOST: ${param:db_host}
    DB_PORT: ${param:db_port}
    DB_PASSWORD: ${param:db_password}
    DB_USERNAME: ${param:db_username}
    LOG_CHANNEL: stderr
    # Serve static assets via CloudFront
    ASSET_URL: //${param:domain}
    LIVEWIRE_ASSET_URL: //${param:domain}
    MIX_ASSET_URL: //${param:domain}
    REDIS_HOST: !GetAtt ElasticCacheCluster.RedisEndpoint.Address
    SESSION_DRIVER: cookie
    VIEW_COMPILED_PATH: /tmp/storage/framework/views

package:
  # Directories to exclude from deployment
  patterns:
    - '!node_modules/**'
    - '!public/storage'
    - '!resources/assets/**'
    - '!storage/**'
    - '!tests/**'

functions:

  # This function runs the Laravel website/API
  web:
    handler: public/index.php
    timeout: 28 # in seconds (API Gateway has a timeout of 29 seconds)
    layers:
      - ${bref:layer.php-81-fpm}
      - ${bref-extra:newrelic-php-81}
        # - ${bref-extra:pgsql-php-81}
      - ${bref-extra:redis-php-81}
    events:
      - httpApi: '*'
    vpc:
      securityGroupIds:
        - !GetAtt ServerlessSecurityGroup.GroupId
        - ${ssm:/laravel/mysql_security_group_id}
      subnetIds:
        - ${param:private_subnet_1a_id}
        - ${param:private_subnet_1b_id}
        - ${param:private_subnet_db1_id}
        - ${param:private_subnet_db2_id}

  # This function lets us run artisan commands in Lambda
  artisan:
    handler: artisan
    timeout: 120 # in seconds
    layers:
      - ${bref:layer.php-81} # PHP
        # - ${bref-extra:pgsql-php-81}
      - ${bref-extra:redis-php-81}
      - ${bref:layer.console} # The "console" layer
    vpc:
      securityGroupIds:
        - !GetAtt ServerlessSecurityGroup.GroupId
        - ${ssm:/laravel/mysql_security_group_id}
      subnetIds:
        - ${param:private_subnet_1a_id}
        - ${param:private_subnet_1b_id}
        - ${param:private_subnet_db1_id}
        - ${param:private_subnet_db2_id}

plugins:
  - ./vendor/bref/bref
  - ./vendor/bref/extra-php-extensions
  - serverless-lift

resources:
  Resources:
    SubnetRouteTableAssociationLambdaPrivateA:
      Type: AWS::EC2::SubnetRouteTableAssociation
      Properties:
        SubnetId: ${param:private_subnet_1a_id}
        RouteTableId: ${param:private_route_table_1a_id}

    SubnetRouteTableAssociationLambdaPrivateB:
      Type: AWS::EC2::SubnetRouteTableAssociation
      Properties:
        SubnetId: ${param:private_subnet_1b_id}
        RouteTableId: ${param:private_route_table_1b_id}

    SubnetRouteTableAssociationLambdaPublicA:
      Type: AWS::EC2::SubnetRouteTableAssociation
      Properties:
        SubnetId: ${param:public_subnet_1a_id}
        RouteTableId: ${param:public_route_table_id}

    ServerlessSecurityGroup:
      Type: AWS::EC2::SecurityGroup
      Properties:
        GroupDescription: SecurityGroup for Serverless Functions
        VpcId: ${param:vpc_id}

    ServerlessRedisSecurityGroup:
      Type: AWS::EC2::SecurityGroup
      Properties:
        GroupDescription: Ingress for Redis Cluster
        VpcId: ${param:vpc_id}
        SecurityGroupIngress:
        - IpProtocol: tcp
          FromPort: '6379'
          ToPort: '6379'
          SourceSecurityGroupId:
            Ref: ServerlessSecurityGroup

    ServerlessCacheSubnetGroup:
      Type: AWS::ElastiCache::SubnetGroup
      Properties:
        Description: "Cache Subnet Group"
        SubnetIds:
          - ${param:private_subnet_1a_id}
          - ${param:private_subnet_1b_id}

    ElasticCacheCluster:
      DependsOn: ServerlessRedisSecurityGroup
      Type: AWS::ElastiCache::CacheCluster
      Properties:
        AutoMinorVersionUpgrade: true
        CacheNodeType: ${param:cache_instance_size}
        CacheSubnetGroupName:
          Ref: ServerlessCacheSubnetGroup
        Engine: redis
        NumCacheNodes: 1
        VpcSecurityGroupIds:
          - !GetAtt ServerlessRedisSecurityGroup.GroupId

constructs:
  website:
    type: server-side-website
    assets:
      # Directories: routes must end with '/*'
      '/css/*': public/css
      '/img/*': public/img
      '/js/*': public/js
      '/vendor/*': public/vendor
      # Files:
      '/favicon.ico': public/favicon.ico
      '/mix-manifest.json': public/mix-manifest.json
      '/robots.txt': public/favicon.ico
    certificate: ${param:certificate_arn}
    domain: ${param:domain}
</pre>
</details>

<details>
<summary>
`.env`
</summary>
<pre>
APP_NAME=FUSE
APP_ENV=local
APP_KEY=base64:deadbeef
APP_DEBUG=true
APP_URL=https://my.example.test

ASSET_URL=https://my.example.test
LIVEWIRE_ASSET_URL=https://my.example.test
MIX_ASSET_URL=https://my.example.test
ASSET_URL=https://my.example.test

LOG_CHANNEL=stack
LOG_LEVEL=debug

DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=example
DB_USERNAME=example
DB_PASSWORD=example

SCOUT_DRIVER=database

BROADCAST_DRIVER=log
CACHE_DRIVER=redis
QUEUE_CONNECTION=sync
SESSION_DRIVER=cookie
SESSION_LIFETIME=120

REDIS_HOST=127.0.0.1
REDIS_PASSWORD=null
REDIS_PORT=6379

MAIL_MAILER=smtp
MAIL_HOST=mailhog
MAIL_PORT=1025
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS=null
MAIL_FROM_NAME="${APP_NAME}"

AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=''
AWS_DEFAULT_REGION=us-east-2
AWS_ENDPOINT='https://s3.us-east-2.amazonaws.com'

# Public
AWS_PUBLIC_BUCKET='example-prod-my-public'
AWS_PUBLIC_URL=''

# Artifacts
AWS_ARTIFACT_ACCESS_KEY_ID=
AWS_ARTIFACT_SECRET_ACCESS_KEY=
AWS_ARTIFACT_REGION=us-east-2
AWS_ARTIFACT_ENDPOINT='https://s3.us-east-2.amazonaws.com'
AWS_ARTIFACT_BUCKET='example-dev-artifacts'
AWS_ARTIFACT_URL=''

FILESTACK_API_KEY=''
FILESTACK_SECURITY_SECRET=''

WISTIA_TOKEN=''

PUSHER_APP_ID=
PUSHER_APP_KEY=
PUSHER_APP_SECRET=
PUSHER_APP_CLUSTER=mt1

MIX_PUSHER_APP_KEY="${PUSHER_APP_KEY}"
MIX_PUSHER_APP_CLUSTER="${PUSHER_APP_CLUSTER}"

</pre>
</details>

**Comment by selfsimilar** [comment]
more supporting details:

<details>
<summary>
composer.json
</summary>
<pre>
{
    "name": "laravel/laravel",
    "type": "project",
    "description": "The Laravel Framework.",
    "keywords": [
        "framework",
        "laravel"
    ],
    "license": "MIT",
    "require": {
        "php": "^8.0",
        "bref/bref": "^1.5",
        "bref/extra-php-extensions": "^0.11.30",
        "bref/laravel-bridge": "^1.2",
        "filestack/filestack-php": "^1.1",
        "fruitcake/laravel-cors": "^2.0",
        "guzzlehttp/guzzle": "^7.0.1",
        "joelbutcher/socialstream": "^3.3",
        "lab404/laravel-impersonate": "^1.7",
        "laravel/framework": "^9.0",
        "laravel/jetstream": "^2.8",
        "laravel/sanctum": "^2.11",
        "laravel/scout": "^9.4",
        "laravel/tinker": "^2.5",
        "league/flysystem-aws-s3-v3": "^3.0",
        "livewire/livewire": "^2.5",
        "packbackbooks/lti-1p3-tool": "^5.1",
        "plank/laravel-mediable": "^5.4",
        "selfsimilar/laravel-d7-password": "^0.1.2",
        "socialiteproviders/generators": "^4.6",
        "socialiteproviders/manager": "^4.1",
        "spatie/laravel-translatable": "^5.1"
    },
    "require-dev": {
        "fakerphp/faker": "^1.9.1",
        "laravel/dusk": "^6.15",
        "laravel/sail": "^1.12",
        "laravel/telescope": "^4.6",
        "mockery/mockery": "^1.4.4",
        "nunomaduro/collision": "^6.0",
        "phpunit/phpunit": "^9.5.10",
        "spatie/laravel-ignition": "^1.0"
    },
    "autoload": {
        "psr-4": {
            "App\\": "app/",
            "Database\\Factories\\": "database/factories/",
            "Database\\Seeders\\": "database/seeders/",
            "SocialiteProviders\\Clever\\": "SocialiteProviders/src/Clever/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "Tests\\": "tests/"
        }
    },
    "scripts": {
        "post-autoload-dump": [
            "Illuminate\\Foundation\\ComposerScripts::postAutoloadDump",
            "@php artisan package:discover --ansi"
        ],
        "post-root-package-install": [
            "@php -r \"file_exists('.env') || copy('.env.example', '.env');\""
        ],
        "post-create-project-cmd": [
            "@php artisan key:generate --ansi"
        ],
        "tlint": [
            "./vendor/bin/tlint"
        ]
    },
    "extra": {
        "laravel": {
            "dont-discover": [
                "laravel/telescope"
            ]
        }
    },
    "config": {
        "optimize-autoloader": true,
        "preferred-install": "dist",
        "sort-packages": true
    },
    "minimum-stability": "dev",
    "prefer-stable": true
}
</pre>
</details>

<details>
<summary>
package.json
</summary>
<pre>
{
    "name": "laravel",
    "version": "1.0.1",
    "repository": {
        "type": "git",
        "url": ""
    },
    "license": "MIT",
    "homepage": "https://my.example.net",
    "private": true,
    "scripts": {
        "dev": "npm run development",
        "development": "mix",
        "watch": "mix watch",
        "watch-poll": "mix watch -- --watch-options-poll=1000",
        "hot": "mix watch --hot",
        "prod": "npm run production",
        "production": "mix --production"
    },
    "devDependencies": {
        "@tailwindcss/forms": "^0.4.0",
        "@tailwindcss/typography": "^0.5.0",
        "alpinejs": "^3.0.6",
        "autoprefixer": "^10.4.0",
        "axios": "^0.21",
        "browser-sync": "^2.27.9",
        "browser-sync-webpack-plugin": "^2.3.0",
        "dotenv": "^10.0.0",
        "laravel-mix": "^6.0.39",
        "lodash": "^4.17.19",
        "postcss": "^8.4.4",
        "postcss-import": "^14.0.1",
        "serverless-lift": "^1.19.0",
        "serverless-s3-sync": "^3.0.0",
        "tailwindcss": "^3.0.0",
        "webpack-cli": "^4.9.2"
    }
}
</pre>
</details>

**Comment by joshhanley** [comment, marked as answer]
@selfsimilar I've not used bref, so can't be of much help, but have a look at this comment and see if that helps https://github.com/livewire/livewire/discussions/4993#discussioncomment-2821628

**Reply by selfsimilar** [reply]
YES! I had a hunch that CloudFront was at fault, but this just saved me several hours of research and trial and error reconfiguration.

For the record, the change in my configuration that fixes it is the following:

```yml
constructs:
  website:
    type: server-side-website
    assets:
      # Directories: routes must end with '/*'
      '/css/*': public/css
      '/img/*': public/img
      '/js/*': public/js
      '/vendor/*': public/vendor
      # Files:
      '/favicon.ico': public/favicon.ico
      '/mix-manifest.json': public/mix-manifest.json
      '/robots.txt': public/favicon.ico
    certificate: ${param:certificate_arn}
    domain: ${param:domain}
    forwardedHeaders:
      - Accept
      - Accept-Language
      - Authorization
      - Content-Type
      - Origin
      - Referer
      - User-Agent
      - X-Forwarded-Host
      - X-Requested-With
      - X-Csrf-Token
      - X-Livewire
```

I'll try to review the Bref/Livewire/Laravel docs to both see where I missed important HTTP Header information, as well as look for where improvements to the docs can be made. As Bref has a dedicated section to Laravel and explicitly makes use of of the serverless-lift plugin, that's where I'll start. But I just double-checked and the `X-Livewire` header is not mentioned once in the Livewire documentation pages for [Deployment](https://laravel-livewire.com/docs/2.x/deployment), [Security](https://laravel-livewire.com/docs/2.x/security), or [Troubleshooting](https://laravel-livewire.com/docs/2.x/troubleshooting).

**Reply by joshhanley** [reply]
@selfsimilar glad it helped! 🙂 PR's are welcome to the docs https://github.com/livewire/docs
